src/algebraic_geometry/EllipticCurve/point.lean

```diff
@@ -0,0 +1,328 @@
+
+include h
+
+/-- The $y$-coordinate of the negation of an affine point. -/
+@[simp] def neg_y : K := -y - ↑E.a₁ * x - ↑E.a₃
+
+/-- The negation of an affine point in `E` lies in `E`. -/
+lemma weierstrass_neg : E.weierstrass x $ neg_y h :=
+by { rw [weierstrass] at h, rw [weierstrass, neg_y, ← h], ring1 }
+
+omit h
+
+/-- Negation in `E⟮K⟯`. -/
+@[simp] def neg : E⟮K⟯ → E⟮K⟯
+| 0            := 0
+| (some x y h) := some x (neg_y h) $ weierstrass_neg h
+
+instance : has_neg E⟮K⟯ := ⟨neg⟩
+
+lemma neg_def (P : E⟮K⟯) : -P = neg P := rfl
+
+@[simp] lemma neg_zero : -0 = (0 : E⟮K⟯) := rfl
+
+@[simp] lemma neg_some : -some x y h = some x (neg_y h) (weierstrass_neg h) := rfl
+
+@[simp] lemma neg_neg (P : E⟮K⟯) : -(-P) = P := by { cases P, { refl }, { simp, ring1 } }
+
+instance : has_involutive_neg E⟮K⟯ := ⟨neg, neg_neg⟩
+
+end negation
+
+/-! ### Doubling in `E⟮K⟯` -/
+
+section doubling
+
+variables {x y : K} (h : E.weierstrass x y)
+
+include h
+
+/-- The gradient of the tangent line at an affine point. -/
+@[simp] def dbl_L : K := (3 * x ^ 2 + 2 * ↑E.a₂ * x + ↑E.a₄ - ↑E.a₁ * y) / (y - neg_y h)
+
+/-- The $x$-coordinate of the doubling of an affine point. -/
+@[simp] def dbl_x : K := dbl_L h ^ 2 + ↑E.a₁ * dbl_L h - ↑E.a₂ - 2 * x
+
+/-- The $y$-coordinate of the doubling of an affine point before applying the final negation. -/
+@[simp] def dbl_y' : K := dbl_L h * (dbl_x h - x) + y
+
+/-- The $y$-coordinate of the doubling of an affine point after applying the final negation. -/
+@[simp] def dbl_y : K := -dbl_y' h - ↑E.a₁ * dbl_x h - ↑E.a₃
+
+/-- The negation of the doubling of an affine point in `E` lies in `E`. -/
+lemma weierstrass_dbl' (hy : y ≠ neg_y h) : E.weierstrass (dbl_x h) (dbl_y' h) :=
+begin
+  rw [weierstrass, ← sub_eq_zero],
+  convert_to
+    dbl_L h * (dbl_L h * (dbl_L h * (y - neg_y h)
+                          + (-3 * x ^ 2 + ↑E.a₁ ^ 2 * x - 2 * ↑E.a₂ * x + 3 * ↑E.a₁ * y
+                              + ↑E.a₁ * ↑E.a₃ - ↑E.a₄))
+                + (-6 * ↑E.a₁ * x ^ 2 - 6 * x * y - 3 * ↑E.a₁ * ↑E.a₂ * x - 3 * ↑E.a₃ * x
+                    + ↑E.a₁ ^ 2 * y - 2 * ↑E.a₂ * y - ↑E.a₁ * ↑E.a₄ - ↑E.a₂ * ↑E.a₃))
+    + (8 * x ^ 3 + 8 * ↑E.a₂ * x ^ 2 - 2 * ↑E.a₁ * x * y + y ^ 2 + 2 * ↑E.a₂ ^ 2 * x + 2 * ↑E.a₄ * x
+        - ↑E.a₁ * ↑E.a₂ * y + ↑E.a₃ * y + ↑E.a₂ * ↑E.a₄ - ↑E.a₆) = 0,
+  { simp only [dbl_x, dbl_y', neg_y], ring1 },
+  field_simp [-neg_y, sub_ne_zero_of_ne hy],
+  rw [weierstrass] at h,
+  linear_combination (2 * y + ↑E.a₁ * x + ↑E.a₃) ^ 2 * h
+    with { normalization_tactic := `[rw [neg_y], ring1] }
+end
+
+/-- The doubling of an affine point in `E` lies in `E`. -/
+lemma weierstrass_dbl (hy : y ≠ neg_y h) : E.weierstrass (dbl_x h) (dbl_y h) :=
+weierstrass_neg $ weierstrass_dbl' h hy
+
+end doubling
+
+/-! ### Addition in `E⟮K⟯` -/
+
+section addition
+
+variables {x₁ y₁ x₂ y₂ : K} (h₁ : E.weierstrass x₁ y₁) (h₂ : E.weierstrass x₂ y₂)
+
+include h₁ h₂
+
+/-- The gradient of the line joining two affine points. -/
```

Comment:
See above. State under which assumption(s) these make sense.

Reply:
I will do this too.